Kube-proxy: Get nodeIPs for both families with dual-stack

uablrek opened the pull request:
#### What type of PR is this?

/kind bug
/sig network
/area kube-proxy



#### What this PR does / why we need it:

For kube-proxy in dual-stack mode the `nodeIP` was only set for the primary family, while the other got an "any" address. This PR uses the nodeIPs in the node object (if possible)

#### Which issue(s) this PR fixes:

Fixes #119524

#### Special notes for your reviewer:

~The "bindAddress" no longer have precedence over the nodeIP in the K8s node object, but is only used as fallback. It is the way it should be IMO, and shouldn't have any backward compatibility issues since the nodeIP is only used at one place in the proxiers (so far).~

The "bindAddress" overrides an address in the node object if it's specified and not the "any" address or a loopback address.

The returned nodeIPs are never an "any" address (zeroes), if the address can't be read from the node object, localhost (127.0.0.1 or ::1) is returned.

The log printouts mentioned in #119524 looks like this with this PR:
```
I0723 17:13:56.934852     360 proxier.go:405] "Record nodeIP and family" nodeIP="192.168.1.2" family="IPv4"
I0723 17:13:56.934997     360 proxier.go:405] "Record nodeIP and family" nodeIP="fd00::c0a8:102" family="IPv6"
```

#### Does this PR introduce a user-facing change?

```release-note
The `--bind-address` parameter in kube-proxy is misleading, no port is opened with this address. Instead it is translated internally to "nodeIP". The nodeIPs for both families are now taken from the Node object if `--bind-address` is unspecified or set to the "any" address (0.0.0.0 or ::). It is recommended to leave `--bind-address` unspecified, and in particular avoid to set it to localhost (127.0.0.1 or ::1)
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

uablrek commented:
I reverted the detectNodeIPs() update since unit tests fails. The unit tests seem to test that with bindAddress=:: and only an ExternalIP (which K8s never sets) the nodeIP should be the ipv4 address, and family IPv4. It looks to me that in dual-stack that would actually result in 2 IPv4 proxiers, or at best an IPv6 proxier with an IPv4 nodeIP.

The single-stack legacy seems a bit too strong here.

uablrek commented:
/retest-required

uablrek commented:
/test pull-kubernetes-e2e-gci-gce-ipvs

sftim commented:
Is this change user-visible? If a cluster operator could ever spot that we've made the change, we should changelog it.

uablrek commented:
@sftim You are right. I will try to figure out what users might experience and write a release note.

uablrek commented:
I have investigated the possible user implact, and for proxy-mode=ipvs there is none. For proxy-mode=iptables there is a potential impact. Here is a proposal:

> In dual-stack clusters, access from a K8s node (or hostNetwork:true) now works for the non-primary family when "loadBalancerSourceRanges" are used

But it works already, _unless_ the loadBalancerIP is also assigned to a local interface. So I am unsure...

And now the _really weird_ corner-case that I really don't want to explain in a release-note:

If the user has defined the kube-proxy `bindAddress` individually for all K8s nodes (i.e. different config for kube-proxy instances) with the purpose to override the (primary-family) nodeIP in the node object to allow access from that address to a loadBalancerIP that is also assigned to a local interface, _that will now fail_.

This since the nodeIP in the node object now have precedence over the bindAddress. I should be noted that defining the bindAddress individually for kube-proxy instance usually is quite hard.

danwinship commented:
Why does this have to change the behavior in the "explicitly-specified `bindAddress`" case? Can't you just leave that broken-ish and only fix the "no `bindAddress`" case?

Also, we could extend `bindAddress` to allow a comma-separated pair of IPs like we've done with some other options (eg `clusterCIDR`). Then people who want to override could just override both families.

(Also, "bind address" is a terrible name (no current proxy backend is actually doing any "binding") and maybe we should deprecate it and replace it with `nodeIP` / `--node-ip`?)

Also, if we're fixing the behavior here, note that the winkernel proxy has its own "autodetect the secondary node IP if it was unspecified" code, which we should be able to remove.

danwinship commented on pkg/util/node/node.go:
```diff
@@ -103,20 +103,10 @@ func GetNodeHostIPs(node *v1.Node) ([]net.IP, error) {
 // If required, it will wait for the node to be created.
-func GetNodeIP(client clientset.Interface, name string) net.IP {
-	var nodeIP net.IP
+func GetNodeIPs(client clientset.Interface, name string) []net.IP {
```
FWIW it looks like `GetNodeIP` is only used by kube-proxy, and it's not obvious that any other part of k/k would have any need for it. (It's only useful for components that are running on a node but aren't kubelet and aren't running in a pod.)

So maybe `GetNodeIPs` should be in `cmd/kube-proxy` or `pkg/proxy/util` and you can submit a separate PR to just remove `GetNodeIP` later? (which saves us needing a sig-node approver on this PR...)

uablrek replied:
Good idea. I'll create a PR to move GetNodeIP. Should I do the GetNodeIP -> GetNodeIPs in that PR? Or keep it as simple as possible since it must be approved by sig-node?

uablrek replied:
Oh, sorry I mis-read. It wasn't a move, but to create a new GetNodeIPs. I'll get to it...

uablrek commented:
> (Also, "bind address" is a terrible name (no current proxy backend is actually doing any "binding") and maybe we should deprecate it and replace it with nodeIP / --node-ip?)

I agree 100% and that's why I changed the default. I have seen people (myself included) that sets bind address to 127.0.0.1 "for security reasons", without knowing about it's use.

> Why does this have to change the behavior in the "explicitly-specified bindAddress" case? Can't you just leave that broken-ish and only fix the "no bindAddress" case?

You are right, a change would probably break some one's use case.

But the reason that I bumped into this problem is that a route with "src" fixes some use-cases in proxy-mode=ipvs, and I wanted to see if I could use the nodeIP. But if many uses set's bind-address to localhost, that can't be done. Also in https://github.com/kubernetes/kubernetes/issues/119656 I suggested to use nodeIP to restrict access. Again localhost isn't useful.

Would it be acceptable to _not_ accept localhost (127.0.0.0/8 or ::1/128) even if explicitly set as bind-address?

Like;

1. If bind-address is specified and not "any" (0.0.0.0, ::) or localhost, then use it.
2. Otherwise use the nodeIPs from the node object
3. If those can't be read, resort to localhost (which will be useless)

uablrek commented:
If a loopback address is specified as bindAddress, it may override the primary family, but the address is still taken from the node object.

IMO to specify a bindAddress of the non-primary family (loopback or not) is a misconfiguration, and probably a mistake. But the current approach is backward compatible (except that the useless loopback address isn't used)

uablrek commented:
By the updated code the weird corner case described in https://github.com/kubernetes/kubernetes/pull/119525#issuecomment-1655421023 is removed.

This PR is now a bug-fix that makes `loadBalancerSourceRanges` work as intended for the non-primary family in a dual-stack cluster.

@sftim Is a release node necessary in this case? Or is it enough to simply refer to solved issues?

uablrek commented:
/retest

danwinship commented on cmd/kube-proxy/app/server.go:
```diff
@@ -904,8 +904,8 @@ func (s *ProxyServer) birthCry() {
 }
 
-// detectNodeIPs returns the proxier's "node IP" or IPs, and the IP family to use if the
-// node turns out to be incapable of dual-stack. (Note that kube-proxy normally runs as
+// detectNodeIPs returns the proxier's "node IPs" and the primary IP family.
+// (Note that kube-proxy normally runs as
```
Old comment is still correct? The node may only have one IP

danwinship commented on cmd/kube-proxy/app/server.go:
```diff
@@ -913,31 +914,84 @@ func (s *ProxyServer) birthCry() {
+//  3. Resort to loopback addresses
 //
 // In all cases, the secondary IP is the zero IP of the other IP family.
 func detectNodeIPs(client clientset.Interface, hostname, bindAddress string) (v1.IPFamily, map[v1.IPFamily]net.IP) {
```
There are a bunch of changes here and I'm still thinking about them...

> Would it be acceptable to _not_ accept localhost (127.0.0.0/8 or ::1/128) even if explicitly set as bind-address?

I don't think so; it doesn't seem totally implausible that someone might want to run a local-only kubernetes install...

It seems like the problem here is just that most of the people saying `--bind-address 127.0.0.1` don't actually want the behavior that they are requesting.

danwinship replied:
> I don't think so; it doesn't seem totally implausible that someone might want to run a local-only kubernetes install...

LOL, no. Now _I'm_ getting confused about what `--bind-address` means.

OK, so the problem, backward-compatibility-wise, is that `--bind-address` _defines_ what `LoadBalancerSourceRanges` value has the magic "allowFromNode" semantics. So if someone is using `--bind-address 127.0.0.1`, then they can specify `LoadBalancerSourceRanges: ["10.0.0.0/8", "127.0.0.1/32"]` and that means "allow from 10.0.0.0/8 and the node". If we started ignoring `--bind-address 127.0.0.1`, then that would change the interpretation of their existing `LoadBalancerSourceRanges`. :exploding_head: 

Is anyone doing that? :man_shrugging:
Are they bad people if they are? Yes!
Are we bad people for designing the API this way? ... No comment.

uablrek replied:
I will verify, but I think you can't send packets 127.0.0.1 -> loadBalancerIP because that would be a "martian". So, that's why I have claimed that localhost addresses in --bind-address (which isn't a bind address) are useless.

uablrek replied:
Verified: including "127.0.0.0/8" in `loadBalancerSourceRanges` _and_ set `bindAddress: "127.0.0.1"` in the kube-proxy config does _not_ allow access from the node, even for proxy-mode=iptables.

<details><summary>Service manifest</summary>

```yaml
apiVersion: v1
kind: Service
metadata:
  name: alpine
spec:
  loadBalancerSourceRanges:
#  - "192.168.1.0/24"
  - "127.0.0.0/8"
  - "192.168.2.0/24"
  - "fd00::c0a8:100/120"
  - "fd00::c0a8:200/120"
  ipFamilyPolicy: RequireDualStack
  selector:
    app: alpine
  type: LoadBalancer
  allocateLoadBalancerNodePorts: false
  ports:
  - port: 5001
    name: nc
```

</details>

```
# In the kube-proxy log
I0803 13:28:06.834076     359 server_others.go:110] "Detected node IP" address="127.0.0.1"
# from a vm on 192.168.2.0/24
nc 10.0.0.0:5001
alpine-6f86fcb8c4-s8t52
# (works)
# On the node:
nc 10.0.0.0:5001
# (hangs...)
# The problem:
ip ro get 10.0.0.0
10.0.0.0 via 192.168.1.201 dev eth1 src 192.168.1.2 uid 0 
# Linux will not pick the loopback as src.
```

uablrek replied:
The "allowFromNode" (hence the entire --bind-address) thing is _only_ useful if the loadBalancerIP is allocated to an interface on the node. So, the above non-working scenario can be "fixed" with;
```
ip add add 10.0.0.0/32 dev eth0
nc 10.0.0.0:5001
alpine-6f86fcb8c4-7fjr4 # (now works)
```

danwinship replied:
> The "allowFromNode" (hence the entire --bind-address) thing is only useful if the loadBalancerIP is allocated to an interface on the node

Well, it's not that it's only _useful_ in that case, it's that it's only _needed_ in that case. If the load balancer IP _isn't_ assigned to a local interface, then node-to-loadbalancer-IP traffic will actually get sent to the load balancer, and then the load balancer can validate the source IP against its own copy of the service's LoadBalancerSourceRanges. (Or, if the LB doesn't implement LoadBalancerSourceRanges itself, then it would either (a) forward the packet to a different node, preserving the source IP, in which case that second node would see that the node's IP was in LoadBalancerSourceRanges and accept it, or (b) hairpin the packet back to the same node again, in which case it would be forced to masquerade it because of hairpinning, in which case the kube-proxy allowFromNode rule would get hit, and the packet would be accepted.)

(So in the (a) case, this only works if the _real_ node IP is in LoadBalancerSourceRanges, and wouldn't work if you used a fake `--bind-address` and corresponding fake LoadBalancerSourceRange.)

aojea replied:
I still not fully process all these scenarios, is there something important we should worry about or just an edge case on some weird configuration?

uablrek commented:
> If we didn't have validation, you could say --bind-address pineapple, and then have a service with LoadBalancerSourceRanges: ["10.0.0.0/8", "pineapple"], and it would work to allow traffic from the local node.

It would set allowFromNode=true, and add the rule but it will not make access from the node work (https://github.com/kubernetes/kubernetes/pull/119525#discussion_r1283224252) unless the loadBalancerIP is assigned to a local interface (https://github.com/kubernetes/kubernetes/pull/119525#discussion_r1283266265)

uablrek commented:
I will let localhost override the nodeIP to ensure backward compatibility. If I go on with my intention to set "src" in proxy-mode=ipvs, I can skip it if nodeIP is localhost.

Anyway, this PR will make it possible to do the right thing.

uablrek commented:
The current semantics of `--bind-address` is horrible, like:
```
--bind-address
  If set, allows access from the node main netns to the loadBalancerIP, if the specified address is
  included in loadBalancerSourceRanges for the service *and* and the loadBalancerIP
  is also assigned to a local interface. 
  This option has no meaning for proxy-mode=ipvs.
```
As mentioned, it doesn't matter what the address is.

sftim commented:
The docs _always_ aim to describe how Kubernetes works now. We actively [avoid](https://kubernetes.io/docs/contribute/style/style-guide/#avoid-statements-about-the-future) documenting how it might work in the future (there are exceptions to the general rule).

uablrek commented:
I agree with Dan, the function is so obscure that is should be deprecated, see https://github.com/kubernetes/kubernetes/pull/119525#discussion_r1283177265, and not described since it only will be confusing, we seasoned kube-proxy maintainers have a hard time wrapping our heads around this :smile:  But IMO it should be deptecated not longer-term, but asap.

sftim commented:
Sorry to be the bearer of bad news but we do at least _aim_ to document behaviors that are present, even deprecated ones.

uablrek commented:
I finally found it: [Hyrum's Law](https://www.hyrumslaw.com/) :smile: 

To actually use --bind-address in meaningful way, the user must have examined the source code _very_ closely. But K8s has many users, so probably someone has, and replacing localhost with a valid nodeIP, which make perfect sense IMHO, will probably ruin someone's day.

danwinship commented:
> The docs _always_ aim to describe how Kubernetes works now. We actively [avoid](https://kubernetes.io/docs/contribute/style/style-guide/#avoid-statements-about-the-future) documenting how it might work in the future (there are exceptions to the general rule).

Right, I don't want to document what it might do in the future, I just want to be clear that it is an "override kube-proxy's idea of what the node IP is" flag, not a "configure LoadBalancerSourceRanges" flag. At the moment, on Linux (but not Windows), overriding kube-proxy's idea of what the node IP is has no effect on anything except LoadBalancerSourceRanges on certain cloud platforms, but that was not the case in the past, and it won't be the case in the future either.

(The existing help text is completely wrong though, and I can see why it leads to people mistakenly doing `--bind-address 127.0.0.1`. So we should do something...)

danwinship commented on cmd/kube-proxy/app/server.go:
```diff
@@ -913,31 +914,84 @@ func (s *ProxyServer) birthCry() {
+			family1 = v1.IPv6Protocol
+		}
+		nodeIPs[family1] = ips[0]
+		primaryFamily = family1
```
It puts both internal and external IPs into the `allIPs` list, but then it constructs [the `nodeIPs` return value](https://github.com/kubernetes/kubernetes/blob/2c6c4566eff972d6c1320b5f8ad795f88c822d09/pkg/util/node/node.go#L95-L101) from the first value in `allIPs`, followed by the first value in `allIPs` of the other IP family, if there is one.

This is consistent with the old single-stack definition of "primary node IP" ("the first InternalIP, unless there are no InternalIPs, in which case the first ExternalIP instead"), and also handles the case of a node that has only InternalIPs for IPv4 and only ExternalIPs for IPv6, which was something we thought would be a common case back before we decided that globally-routable IPv6 IPs could still be called InternalIPs.

danwinship replied:
... I feel like there used to be a comment on one of those functions explaining the definition of primary and secondary node IP, but there isn't currently...

uablrek replied:
I missed the allIPs/nodeIPs thing, sorry. I removed the check, but made a comment.

> I feel like there used to be a comment on one of those functions explaining the definition of primary and secondary node IP, but there isn't currently...

Yeah, and please note that what I personally consider the primary nodeIP, i.e. the ip of the family that is used for services that doesn't explicitly specify a family, can be overridden by specifying a --bind-address of the _other_ family.

uablrek committed:
Kube-proxy: handle dual-stack in detectNodeIPs()

uablrek commented:
Updated after review:
* Removed logging
* Allow override with localhost
* Don't check that nodeIPs are of different families, but add a comment
* Keep as much as possible of the "order of precedence" comment
* Only one commit since everything is in one place now (except tests)

uablrek commented:
I suggest that the --bind-address option is obsoleted without further comments, and that the extra iptables rule added when allowFromNode==true is instead _always_ added when LoadBalancerSourceRanges are specified. The extra rule doesn't do any harm, and LoadBalancerSourceRanges are probably not used to the extend that the extra rule becomes a burden.

uablrek commented:
/retest

aojea commented:
I knew this discussion sounded familiar XD https://github.com/kubernetes/kubernetes/issues/95515

aojea commented on cmd/kube-proxy/app/server.go:
```diff
@@ -941,3 +944,33 @@ func detectNodeIPs(client clientset.Interface, hostname, bindAddress string) (v1
 	}
 }
+
+// getNodeIP returns IPs for the node with the provided name.  If
+// required, it will wait for the node to be created.
+func GetNodeIPs(client clientset.Interface, name string) []net.IP {
```
do we want to export this function? why don't we do the other way around and make this function private?
```suggestion
// getNodeIP returns IPs for the node with the provided name.  If
// required, it will wait for the node to be created.
func getNodeIPs(client clientset.Interface, name string) []net.IP {
```

uablrek replied:
Yes, it's used in some "hollow proxier"

aojea commented on cmd/kube-proxy/app/server.go:
```diff
@@ -914,30 +915,75 @@ func (s *ProxyServer) birthCry() {
-		klog.InfoS("Can't determine this node's IP, assuming 127.0.0.1; if this is incorrect, please set the --bind-address flag")
-		nodeIP = netutils.ParseIPSloppy("127.0.0.1")
+
+	if ips := GetNodeIPs(client, hostname); len(ips) > 0 {
```
before we only did a call to the apiserver if the bindaddress as unspecified, we are doing this call always now and this will add some delay on the bootrap (unless you use LocalModeNodeCIDR that you have to wait for the node nevertheless)

Should we worry about this?

danwinship replied:
Yeah, I noticed this when I reviewed it before. But kube-proxy doesn't actually do anything useful until it's able to talk to the apiserver anyway.

One possible catch: the healthz server doesn't start until after this point, so this will delay kube-proxy readiness if you have a readinessProbe on it...

danwinship commented:
The [comment above](https://github.com/kubernetes/kubernetes/pull/119525#discussion_r1283177265) about hacking it to not change the winkernel behavior still applies.

(Also the comment about `GetNodeIPs` having a comment that calls it `getNodeIPs`.)

Other than that this looks good, unless we're worried about the readinessProbe thing...

uablrek commented:
> OK, long(er)-term, we need to deprecate --bind-address and replace it with something like --override-node-ip, and make it very clear what it is and isn't used for.

I don't think this can be explained in an understandable way, even for a K8s user with great networking experience (see e.g. https://github.com/kubernetes/kubernetes/pull/119525#discussion_r1283370824). What do you think about:

> I suggest that the --bind-address option is obsoleted without further comments, and that the extra iptables rule added when allowFromNode==true is instead always added when LoadBalancerSourceRanges are specified. The extra rule doesn't do any harm, and LoadBalancerSourceRanges are probably not used to the extend that the extra rule becomes a burden.

as suggested above, and sweep the whole thing under the carpet?

And release note something like:
```
The --bind-address has no longer any function and should not be specified. The option is obsoleted and will be removed in a
future version.
```

I can do that, but in another PR since it's an API change.

uablrek commented:
I don't understand if this is a failed test or not:

> pull-kubernetes-e2e-capz-windows — Context retired without replacement.

The robot says it's failed, but it's ticked as green and has no "Details" button.

aojea commented:
> 

that job is not important

aojea commented:
> > 
> 
> that job is not important

Oh, it may be important since we are modifying windoes here

aojea commented on cmd/kube-proxy/app/server.go:
```diff
@@ -914,30 +915,75 @@ func (s *ProxyServer) birthCry() {
-	if nodeIP.IsUnspecified() {
-		nodeIP = utilnode.GetNodeIP(client, hostname)
+	primaryFamily := v1.IPv4Protocol
+	nodeIPs := map[v1.IPFamily]net.IP{
```
is not better to structure the code same as the comment?

```
//  1. if bindAddress is not 0.0.0.0 or ::, then it is used as the primary IP.
//  2. if the Node object can be fetched, then its address(es) is/are used
//  3. otherwise the node IPs are 127.0.0.1 and ::1
```

that way is straightforward to read

danwinship replied:
I'm ok with the current organization...

danwinship commented:
can you squash it all together?

uablrek commented:
> can you squash it all together?

@danwinship Done

aojea commented on cmd/kube-proxy/app/server.go:
```diff
@@ -955,30 +956,75 @@ func (s *ProxyServer) birthCry() {
+//  2. if the Node object can be fetched, then its address(es) is/are used
+//  3. otherwise the node IPs are 127.0.0.1 and ::1
 //
 // In all cases, the secondary IP is the zero IP of the other IP family.
```
this comment needs to be updated then

aojea commented on cmd/kube-proxy/app/server.go:
```diff
@@ -955,30 +956,75 @@ func (s *ProxyServer) birthCry() {
+		nodeIPs, err = utilnode.GetNodeHostIPs(node)
+		if err != nil {
+			klog.ErrorS(err, "Failed to retrieve node IPs")
+			return false, err
```
I think we should not error here and return nil to  keep retrying

https://github.com/kubernetes/kubernetes/blob/49768134e527c1cdc93152f1c575062fb836c08e/pkg/util/node/node.go#L92

It is perfectly possible that the Node starts and the addresses are still not populated
```suggestion
			return false, nil
```

aojea commented:
one last comment before merge https://github.com/kubernetes/kubernetes/pull/119525/files#r1320767795 , I think we should keep retrying there, the rest LGTM too

uablrek committed:
Updates

uablrek commented:
I added the update in https://github.com/kubernetes/kubernetes/pull/119525#discussion_r1320767795, and removed the comment:
> // In all cases, the secondary IP is the zero IP of the other IP family.

But I still don't understand why the windows tests fails. Even if they are not required, this PR will make the always fail if merged, and cause a lot of worries. Right?

I can't test on Windows, but it seems like a code check rather than a runtime bug. So perhaps it can be executed on Linux?

aojea commented:
> I checked again, and it seem to be a timeout. A difference is that now the Node objects are _always_ read. Is it possible that can't be done on Windows at this time?

the job is failing in other PRs too with the same error https://prow.k8s.io/job-history/gs/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-e2e-capz-windows-master and I can't see any error on the kube-proxy logs https://gcsweb.k8s.io/gcs/kubernetes-jenkins/pr-logs/pull/120483/pull-kubernetes-e2e-capz-windows-master/1700103308384931840/artifacts/clusters/capz-conf-p2ijn7/kube-system/

/lgtm

@danwinship for final approval

aojea commented:
/label tide/merge-method-squash

danwinship commented:
/approve

danwinship commented:
@uablrek you should update the release note text though

uablrek commented:
It is hard to describe the **user-facing** aspects IMO. I made an attempt:

> The `--bind-address` parameter in kube-proxy is misleading, no port is opened with this address. Instead it is translated
> internally to "nodeIP". The nodeIPs for both families are now taken from the Node object if `--bind-address` is unspecified
> or set to the "any" address (0.0.0.0 or ::). It is recommended to leave `--bind-address` unspecified, and in particular
> avoid to set it to localhost (127.0.0.1 or ::1)
